src/Compilers/CSharp/Portable/Lowering/IteratorRewriter/IteratorRewriter.cs

```diff
@@ -74,7 +75,91 @@ internal partial class IteratorRewriter : StateMachineRewriter
 
             stateMachineType = new IteratorStateMachine(slotAllocatorOpt, compilationState, method, methodOrdinal, isEnumerable, elementType);
             compilationState.ModuleBuilderOpt.CompilationState.SetStateMachineType(method, stateMachineType);
-            return new IteratorRewriter(body, method, isEnumerable, stateMachineType, slotAllocatorOpt, compilationState, diagnostics).Rewrite();
+            var rewriter = new IteratorRewriter(body, method, isEnumerable, stateMachineType, slotAllocatorOpt, compilationState, diagnostics);
+            if (rewriter.EnsureAllSymbolsAndSignature())
+            {
+                return body;
+            }
+
+            return rewriter.Rewrite();
+        }
+
+        // returns true if any type we depend on is not present
+        protected bool EnsureAllSymbolsAndSignature()
+        {
+            bool hasErrors = false;
+
+            EnsureSpecialType(SpecialType.System_Int32, ref hasErrors);
+            EnsureSpecialMethod(SpecialMember.System_IDisposable__Dispose, ref hasErrors);
+
+            // IEnumerator<T>
+            EnsureSpecialPropertyAsMember(SpecialMember.System_Collections_Generic_IEnumerator_T__Current,
+                                            SpecialType.System_Collections_Generic_IEnumerator_T, ref hasErrors);
```

Comment:
Yes, the `GenerateEnumeratorImplementation` method always gets this type and uses it, and that method is always called when rewriting. 

As far as I can tell, that is also true on the VB side, but the VB `EnsureAllSymbolsAndSignature` has some conditional checks that didn't make sense to me. I'll stop by to discuss.